### PR TITLE
Add hostname options to Keycloak CR

### LINKED ISSUE #14395: Add hostname options to Keycloak CR
Mapped options:
* hostname
* hostname-admin
* hostname-port
* hostname-strict
* hostname-strict-backchannel
* hostname-url
* hostname-admin-url

Omitting for now for simplicity:
* hostname-path (superseded by hostname-url)

Removed/moved existing properties in the CR:
* hostname (moved under hostname/server)
* Remove the [`INSECURE-DISABLE` functionality](https://www.keycloak.org/operator/advanced-configuration#_disabling_required_cr_fields) to align with the distribution

CR structure:
```yaml
apiVersion: k8s.keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
spec:
  hostname:
    hostname: some-server-hostname # moved from spec/hostname; now optional; if set it has to be reflected by Ingress
    hostnameUrl: url
    admin: admin.some-serverhostname # not reflected by Ingress for now
    adminUrl: url
    port: 80 # integer; reflected by Ingress
    strict: false # boolean
    strictBackchannel: false # boolean
...
```

**NOTE:** this is a breaking change and there will be no automated migration; make sure to mention it in release notes and migration guide.

### PR DESCRIPTION
**Pull request by mabartos**
Closes #14395

### CODE REVIEW

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentConfig.java
Comment (mabartos) [inline operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentConfig.java:159]: @vmuzikar In the description of the associated issue is mentioned that the `INSECURE-DISABLE` functionality should be removed. Does it mean that we're gonna support only the production-ready environment with Operator? Will there be no support from the developer's perspective?

Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentConfig.java:159]: jFTR – this was discussed offline and will be reflected in docs.

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java:65]: I think, in theory (a future dev mode or something) this can actually be `null`, right?

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/HostnameSpec.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/HostnameSpec.java:29]: The `hostname` can now be `null` and is therefore optional (in case someone e.g. sets `hostname-strict=false`).

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngress.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngress.java:86]: Comment unrelated to this line. I believe we should not set this if `hostname == null`.

**Review by vmuzikar — APPROVED** [review]

**Review by vmuzikar — CHANGES_REQUESTED** [review]
Sorry, wrong button. :)

**Comment by mabartos** [comment]
@vmuzikar Thanks for your review. I addressed your comments, and I'd like to try if the tests will pass in GHAs. I also need to make changes to the guides.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@mabartos Thanks for the update!

Just a small heads up. There will be some conflicts with this: https://github.com/keycloak/keycloak/pull/14896

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java
Comment (mabartos) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java:111]: Oh, you're right.

**Comment by mabartos** [comment]
@vmuzikar Thanks! I'll update it, but it'd be good to wait until the HTTP opts are merged.

**Comment by vmuzikar** [comment]
@mabartos #14896 was merged, I think we're now ready for a rebase.

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java
Comment (pedroigor) [inline operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java:109]: I'm not sure if we can so strict about the `hostname-port` because you might have proxy in front of the cluster. I'm saying that based on past issues where there is a proxy chain (e.g.: https://github.com/keycloak/keycloak/issues/13261) and people exposing the server using a port other than the defaults.

For me we should either remove `hostname-port` from the CR or remove this validation.

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java
Comment (pedroigor) [inline operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java:110]: We can't assume Nginx controller is the default.

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngress.java
Comment (pedroigor) [inline operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngress.java:100]: I also agree that setting the `hostname` is enough most of the time.

But we also have the `hostname-url` option to force a base URL to frontend URLs. That said, I'm not sure if you want to resolve the host from the `hostname-url` too in case this option is set.

**Review by pedroigor — COMMENTED** [review]
Out of curiosity, the ingress host is set based on the hostname options, right?

If not set, what is the ingress host?

**Comment by vmuzikar** [comment]
@pedroigor Thank you for your input! This made me wondering. Should we then for now omit the hostname-port and hostname-path from the CRD? It'd still be possible to use them via the serverConfiguration free-form field.

**Comment by mabartos** [comment]
@vmuzikar @pedroigor Thanks guys for your comments. 

> If not set, what is the ingress host?
AFAIK, only some external IP is used. But TBH, I'm not sure. I can verify it. Good point!

As is described in the doc for [Ingress rules](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-rules):
_An optional host. In this example, no host is specified, so the rule applies to all inbound HTTP traffic through the IP address specified. If a host is provided (for example, foo.bar.com), the rules apply to that host._

I agree that we could omit the `hostname-port` ATM, as you've suggested, and we can discuss it later as it won't be a breaking change afterward.

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java
Comment (mabartos) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java:109]: Interesting issue with the proxy chain. Thanks for bringing this up. I'm in to remove `hostname-port` for now, as described below.

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngress.java
Comment (mabartos) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngress.java:100]: Good point. 

I'd say we could parse the hostname from the `hostname-url`, if the `hostname` is not specified. Otherwise, prefer only the `hostname`. 

If both are specified, it's the responsibility of the user/dev to manage them, right?!

**Review by vmuzikar — APPROVED** [review]
@mabartos LGTM, thank you!

Let's wait for the tests.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@mabartos Unfortunately, there seem to be some test errors related to this PR. Could you please check it?

**Comment by mabartos** [comment]
@vmuzikar Model test failures are not related to this PR.

**Review by vmuzikar — APPROVED** [review]

**Pull request merged**
